ethcore/src/engines/clique/mod.rs

```diff
@@ -0,0 +1,670 @@
+	block_state_by_hash: RwLock<LruCache<H256, CliqueBlockState>>,
+	proposals: RwLock<HashMap<Address, VoteType>>,
+	signer: RwLock<Option<Box<EngineSigner>>>,
+	step_service: Option<Arc<StepService>>,
+}
+
+impl Clique {
+	/// Initialize Clique engine from empty state.
+	pub fn new(our_params: CliqueParams, machine: EthereumMachine) -> Result<Arc<Self>, Error> {
+		let mut engine = Clique {
+			epoch_length: our_params.epoch,
+			period: our_params.period,
+			client: RwLock::new(Option::default()),
+			block_state_by_hash: RwLock::new(LruCache::new(STATE_CACHE_NUM)),
+			proposals: RwLock::new(Default::default()),
+			signer: RwLock::new(Default::default()),
+			machine,
+			step_service: None,
+		};
+
+		let res =  Arc::new(engine);
+
+		if our_params.period > 0 {
+			engine.step_service = Some(StepService::start(Arc::downgrade(&res) as Weak<Engine<_>>));
+		}
+
+		Ok(res)
+	}
+
+	fn sign_header(&self, header: &Header) -> Result<(Signature, H256), Error> {
+		match self.signer.read().as_ref() {
+			None => {
+				return Err(Box::new("sign_header: No signer available.").into());
+			}
+			Some(signer) => {
+				let digest = header.hash();
+				match signer.sign(digest) {
+					Ok(sig) => {
+						return Ok((sig, digest));
+					}
+					Err(e) => {
+						return Err(Box::new(format!("sign_header: failed to sign header, error: {}", e)).into());
+					}
+				}
+			}
+		}
+	}
+
+	/// Construct an new state from given checkpoint header.
+	fn new_checkpoint_state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		assert_eq!(header.number() % self.epoch_length, 0);
+
+		let mut state = CliqueBlockState::new(
+			match header.number() {
+				0 => NULL_AUTHOR,
+				_ => recover_creator(header)?,
+			},
+			extract_signers(header)?);
+
+		state.calc_next_timestamp(header, self.period);
+
+		Ok(state)
+	}
+
+	fn state_no_backfill(&self, hash: &H256) -> Option<CliqueBlockState> {
+		return self.block_state_by_hash.write().get_mut(hash).cloned();
+	}
+
+	/// get CliqueBlockState for given header, backfill from last checkpoint if needed.
+	fn state(&self, header: &Header) -> Result<CliqueBlockState, Error> {
+		let mut block_state_by_hash = self.block_state_by_hash.write();
+		if let Some(state) = block_state_by_hash.get_mut(&header.hash()) {
+			return Ok(state.clone());
+		}
+		// If we are looking for an checkpoint block state, we can directly reconstruct it.
+		if header.number() % self.epoch_length == 0 {
+			let state = self.new_checkpoint_state(header)?;
+			block_state_by_hash.insert(header.hash().clone(), state.clone());
+			return Ok(state);
+		}
+		// BlockState is not found in memory, which means we need to reconstruct state from last checkpoint.
+		match self.client.read().as_ref().and_then(|w| { w.upgrade() }) {
+			None => {
+				return Err(From::from("failed to upgrade client reference"));
+			}
+			Some(c) => {
+				let last_checkpoint_number = (header.number() / self.epoch_length as u64) * self.epoch_length;
```

Comment:
Needless `div` and `mul` the same value, remove ?!

There is nothing in  `chainspec deserialization` preventing `epoch_length` to be zero i.e., potential `div by 0` (but that should be removed if you accept my first point)

Added some comments in `chainspec mod` there also you use [NonZeroU64](https://doc.rust-lang.org/beta/std/num/struct.NonZeroU64.html) as epoch_length